### PR TITLE
Adaptive BAL

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
This PR adds a few features:
- adaptive BAL which selects up to a minimum inter sample distance instead of a fixed number of samples.
- fixed force features for BAL
- added full gradient features with random projections to the available feature maps
- new visualizations to the BAL IPS node

### CODE REVIEW

**Comment by M-R-Schaefer** [comment]
pre-commit.ci autofix

**Review by Tetracarbonylnickel — CHANGES_REQUESTED** [review]

File: apax/bal/feature_maps.py [hunk 1]
```diff
@@ -110,34 +133,90 @@ def inner(ll_params):
...
+                (gb, gw), _ = jax.tree_util.tree_flatten(g_summed)
                 g = [gw, gb]
                 g = jnp.concatenate(g)
 
+            elif self.strategy == "flatten":
+                g_flat = jax.tree_map(lambda arr: jnp.reshape(arr, (-1,)), g_ll)
+                (gb, gw), _ = jax.tree_util.tree_flatten(g_flat)
+                g = gw
+            else:
+                raise ValueError(f"unknown strategy: {self.strategy}")
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/bal/feature_maps.py:161]: the implementations of the different options is 19 lines of code, I don't think further abstraction is required.

**Review by Tetracarbonylnickel — APPROVED** [review]
